File: 908-middle-of-the-linked-list/middle-of-the-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head.next is None:
            return head
        if head.next.next is None:
            return head.next
        flag = 0
        c = head
        d = head.next
        while d.next is not None:
            c = c.next
            d = d.next
            if d.next is None:
                flag = 1
                break
            d = d.next
        
        if flag == 1:
            return c
        else:
            return c.next
        

File: 908-middle-of-the-linked-list/middle-of-the-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        if(head == nullptr || head->next == nullptr) return head;

        ListNode *temp = new ListNode(0, head);
        ListNode *fast = temp, *slow = temp;
        while(fast != nullptr) {
            cout << slow->val << endl;
            if(fast->next == nullptr) return slow->next;
            slow = slow->next;
            fast = fast->next->next;
        } 

        return slow;
    }
};